File: backend/src/main/java/com/xiojuandawt/blood4life/services/NotificationService.java
package com.xiojuandawt.blood4life.services;

import com.xiojuandawt.blood4life.entities.BloodDonor;
import com.xiojuandawt.blood4life.entities.Notification;

import java.util.List;
import java.util.Optional;

import com.xiojuandawt.blood4life.entities.Hospital;

public interface NotificationService {

    // Get all notifications for a specific blood donor
    List<Notification> getNotificationsByDonor(BloodDonor donor);

    // Get unread notifications count for a specific blood donor
    Long getUnreadCount(BloodDonor donor);

    // Get unread notifications for a specific blood donor
    List<Notification> getUnreadNotifications(BloodDonor donor);

    // Get all notifications for a specific hospital
    List<Notification> getNotificationsByHospital(Hospital hospital);

    // Get unread notifications count for a specific hospital
    Long getUnreadCount(Hospital hospital);

    // Get unread notifications for a specific hospital
    List<Notification> getUnreadNotifications(Hospital hospital);

    // Mark a notification as read
    Optional<Notification> markAsRead(Integer notificationId);

    // Mark all notifications as read for a specific donor
    void markAllAsRead(BloodDonor donor);

    // Mark all notifications as read for a specific hospital
    void markAllAsRead(Hospital hospital);

    // Create a new notification for donor
    Notification createNotification(BloodDonor receiver, String message);

    // Create a new notification for hospital
    Notification createNotification(Hospital receiver, String message);
}

File: frontend/src/services/notificationService.ts
import axiosInstance from '../utils/axiosInstance';

export interface Notification {
    id: number;
    message: string;
    dateNotification: string;
    read: boolean;
}

export interface UnreadCountResponse {
    count: number;
}

/**
 * Notification Service
 * Handles all notification-related API calls
 */
const notificationService = {
    /**
     * Get all notifications for the authenticated user
     */
    getMyNotifications: async (): Promise<Notification[]> => {
        const response = await axiosInstance.get<Notification[]>('/notifications');
        return response.data;
    },

    /**
     * Get the count of unread notifications
     */
    getUnreadCount: async (): Promise<number> => {
        const response = await axiosInstance.get<UnreadCountResponse>('/notifications/unread/count');
        return response.data.count;
    },

    /**
     * Get all unread notifications
     */
    getUnreadNotifications: async (): Promise<Notification[]> => {
        const response = await axiosInstance.get<Notification[]>('/notifications/unread');
        return response.data;
    },

    /**
     * Mark a specific notification as read
     */
    markAsRead: async (id: number): Promise<void> => {
        await axiosInstance.put(`/notifications/${id}/read`);
    },

    /**
     * Mark all notifications as read
     */
    markAllAsRead: async (): Promise<void> => {
        await axiosInstance.put('/notifications/read-all');
    }
};

export default notificationService;
